keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,259 @@
+   * Queues that version for approval if it exists.
+   */
+  fun checkEnvironment(
+    envInfo: EnvironmentInfo
+  ) {
+    val pendingVersionsToCheck: MutableSet<String> =
+      when (envInfo.environment.constraints.anyStateful) {
+        true -> repository
+          .pendingConstraintVersionsFor(envInfo.deliveryConfig.name, envInfo.environment.name)
+          .filter { envInfo.versions.contains(it) }
+          .toMutableSet()
+        false -> mutableSetOf()
+      }
+
+    checkConstraints(
+      envInfo,
+      pendingVersionsToCheck
+    )
+
+    /*
+     * If there are pending constraints for prior versions, that we didn't recheck in the process of
+     * finding the latest version above, recheck in ascending version order
+     * so they can be timed out, failed, or approved.
+     */
+    dealWithOldVersions(envInfo, pendingVersionsToCheck)
+  }
+
+  /**
+   * Looks at all versions for an environment, and determines the latest version that passes constraints,
+   * or null if none pass.
+   *
+   * In the process we check and evaluate constraints for each version.
+   *
+   * If a version passes all constraints it is queued for approval.
+   */
+  private fun checkConstraints(
+    envInfo: EnvironmentInfo,
+    pendingVersionsToCheck: MutableSet<String>
+  ) {
+    var version: String? = null
+    var versionIsPending = false
+    val vetoedVersions: Set<String> =
+      (envInfo.vetoedArtifacts[envPinKey(envInfo.environment.name, envInfo.artifact)]?.versions)
+        ?: emptySet()
+
+    if (envInfo.environment.constraints.isEmpty() && implicitConstraints.isEmpty()) {
+      version = envInfo.versions.firstOrNull { v ->
+        !vetoedVersions.contains(v)
+      }
+    } else {
+      version = envInfo.versions
+        .firstOrNull { v ->
+          pendingVersionsToCheck.remove(v) // remove to indicate we are rechecking this version
+          if (vetoedVersions.contains(v)) {
+            false
+          } else {
+            /**
+             * Only check stateful evaluators if all stateless evaluators pass. We don't
+             * want to request judgement or deploy a canary for artifacts that aren't
+             * deployed to a required environment or outside of an allowed time.
+             */
+            val passesConstraints =
+              checkStatelessConstraints(envInfo.artifact, envInfo.deliveryConfig, v, envInfo.environment) &&
+                checkStatefulConstraints(envInfo.artifact, envInfo.deliveryConfig, v, envInfo.environment)
+
+            if (envInfo.environment.constraints.anyStateful) {
+              versionIsPending = repository
+                .constraintStateFor(envInfo.deliveryConfig.name, envInfo.environment.name, v)
+                .any { it.status == PENDING }
+            }
+
+            // select either the first version that passes all constraints,
+            // or the first version where stateful constraints are pending.
+            passesConstraints || versionIsPending
```

Comment:
Does this actually need to return a version? I thought this was just running constraints and anther component would determine which version is approved.

Reply:
The function is not returning a version, it's just setting it as a variable. On line 123 we queue the version for approval, indicating that it has passed a ll the constraints. 

The `EnvironmentPromotionChecker` reads from that queue